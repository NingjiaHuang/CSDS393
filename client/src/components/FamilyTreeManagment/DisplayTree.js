import React, {Fragment, useState, useEffect} from 'react'
import mock_data from "./MOCK_DATA_TREE.json"

const DisplayTree = (cat) =>{
  const[searchTerm, setSearchTerm] = useState("")
  const[tree, setTree] = useState([]);
  const getTree = async() => {
      try{
          const response = await fetch(`http://localhost:4020/api/v1/gettree/${cat.certi_num}`) //should be get tree of a certain cat
          const jsonData = await response.json()
          setTree(jsonData);
      } catch (err){
          console.error(err.message);
      }
  }

    useEffect(() =>{
        getTree();
    }, [])

    return(
        <Fragment>
             {" "}
          <input type="text" className="form-control" 
          placeholder="search a family member here" 
          onChange={event => {setSearchTerm(event.target.value)}}
          />
          <table class="table mt-5 text-center">
              <thead>
                  <tr>
                      <th>Cat Name</th>
                      <th>Certificate number</th>
                      <th>Sire path</th>
                      <th>Dam path</th>
                  </tr>
              </thead>
              <tbody>
              {mock_data.filter((val) =>{  //if get all tree completed, change "mock_data" to tree
                  if (searchTerm == ""){
                    return val
                  } else if (val.cat_name.toLowerCase().includes(searchTerm.toLowerCase()) ){
                    return val
                  }
              }
              ).map((val, key) => (
                  <tr key={key}>
                      <td>{val.cat_id}</td>
                      <td>{val.cat_name}</td>
                      <td>{val.sire_path}</td>
                      <td>{val.dam_path}</td>
                      <td>
                      </td>
                  </tr>
              ))
              }
            </tbody>
        </table>
        </Fragment>
    )
}

export default DisplayTree;