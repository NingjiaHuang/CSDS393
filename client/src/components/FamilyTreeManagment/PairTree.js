import React, {useState} from 'react'
import { Button, Modal } from 'react-bootstrap';

const PairTree = () =>{
    const [info, setInfo] = useState({
        id:"",
        cat_name:"",
        sire_id:"",
        dam_id:""
    });

    const {id, cat_name, sire_id, dam_id} = info;
        
    const onChange = (e) => {
        setInfo({...info, [e.target.name]:e.target.value})
    }

    const [show, setShow] = useState(false);
    const handleClose = () => setShow(false);
    const handleShow = () => setShow(true);

    const onSubmitForm = async(e) => {
        e.preventDefault();
        try{
            const body = {id, cat_name, sire_id, dam_id};
            console.log(body);
            const response = await fetch("http://localhost:4020/api/v1/cats/tree",{ // url needs update. need to add to breeding cat table
                method:"POST",
                headers: {"Content-Type": "application/json"},
                body: JSON.stringify(body)
            });

            window.location="/family_tree";
        } catch(err){
            console.error(err.message);
        }
    }



    return(
        <div>
                <Button data-testid="pairBtn" variant="info" onClick={handleShow}>Pair</Button>
                <Modal show={show} onHide={handleClose}>
                    <Modal.Header closeButton>
                    <Modal.Title>Pair a kitten with parents</Modal.Title>
                    </Modal.Header>
                    <Modal.Body>
                        <form onSubmit={onSubmitForm}>
                            <label htmlFor='id'>Kitten certificate number: </label>
                            <input type="text" 
                            name="id"   
                            id='id' 
                            className="form-control my-2"
                            onChange={e => onChange(e)}/>

                            <label>Kitten name: </label>
                            <input 
                            type="text" 
                            name="cat_name"  
                            className="form-control my-2"
                            onChange={e => onChange(e)}/>

                            <label>Sire certificate number: </label>
                            <input 
                            type="text" 
                            name="sire_id"  
                            className="form-control my-2"
                            onChange={e => onChange(e)}/>

                            <label>Dam certificate number: </label>
                            <input 
                            type="text" 
                            name="dam_id"   
                            className="form-control my-2"
                            onChange={e => onChange(e)}/>
                        </form>
                    </Modal.Body>
                    <Modal.Footer>
                    <Button variant="secondary" onClick={handleClose}>
                        Close
                    </Button>
                    <Button variant="warning" onClick={onSubmitForm}>
                        Pair
                    </Button>
                    </Modal.Footer>
                </Modal>
        </div>
    )
}


export default PairTree;