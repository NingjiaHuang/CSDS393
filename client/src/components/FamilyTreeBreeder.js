import React from 'react'
import DisplayTree from "./FamilyTreeManagment/DisplayTree"
import PairTree from "./FamilyTreeManagment/PairTree"
import { MDBContainer } from 'mdbreact';

function FamilyTreeBreeder() {
  return(
    <MDBContainer>
     <div className='mx-auto mt-5'>
          <h2 className="text-center">Family Tree</h2>
            <PairTree />
          <div className='mt-3'>
            <DisplayTree />
          </div>
        </div>
  </MDBContainer>

  ); 
}

export default FamilyTreeBreeder;
