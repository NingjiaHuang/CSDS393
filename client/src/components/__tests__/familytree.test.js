import {render, screen, fireEvent, cleanup} from '@testing-library/react';
import DisplayTree from '../FamilyTreeManagment/DisplayTree';
import PairTree from '../FamilyTreeManagment/PairTree';
import FamilyTreeBreeder from '../FamilyTreeBreeder';
import FamilyTreePP from '../FamilyTreePP';
import renderer from 'react-test-renderer'
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure  } from 'enzyme';
import '@testing-library/jest-dom/extend-expect';

configure({ adapter: new Adapter() })
afterEach(() =>{
    cleanup();
})

test('matches snapshot of family tree of parents', () =>{
    const tree = renderer.create(<FamilyTreePP/>).toJSON();
    expect(tree).toMatchSnapshot();
})

test('matches snapshot of family tree of breeder', () =>{
    const tree = renderer.create(<FamilyTreeBreeder/>).toJSON();
    expect(tree).toMatchSnapshot();
})


test('matches snapshot of pair cats', () =>{
    const tree = renderer.create(<PairTree/>).toJSON();
    expect(tree).toMatchSnapshot();
})

test('matches snapshot of display tree', () =>{
    const tree = renderer.create(<DisplayTree/>).toJSON();
    expect(tree).toMatchSnapshot();
})

test('click pair button ', () => {
    const mockOnClick = jest.fn()
    const { getByTestId } = render(<PairTree onClick={mockOnClick()} />)
    const clickIndicator = getByTestId('pairBtn')
    fireEvent.click(clickIndicator)
    expect(mockOnClick).toHaveBeenCalledTimes(1)
})

test('pair tree should render react-modal', () => {
    const wrapper = shallow(<PairTree />);
    expect(wrapper.find('Modal')).toHaveLength(1);
  });

  test("search box should accept text", () => {
    const {getByTestId} = render(<DisplayTree/>);
    const searchInputNode = getByTestId('searchTree');
    expect(searchInputNode.value).toMatch("");
    fireEvent.change(searchInputNode, {target:{value:'Ja'}});
    expect (searchInputNode.value).toMatch("Ja")
})

test('click pair button ', () => {
    const mockOnClick = jest.fn()
    const mockonsubmit = jest.fn()
    const { getByTestId, getByLabelText } = render(<PairTree onClick={mockOnClick()} onsubmit={mockonsubmit}/>)
    const clickIndicator = getByTestId('pairBtn')
    //open the modal
    fireEvent.click(clickIndicator)
    expect(mockOnClick).toHaveBeenCalledTimes(1)

    //edit the modal form
    const id = getByLabelText("Kitten certificate number:");
    const name = getByLabelText("Kitten name:")
    const scert = getByLabelText("Sire certificate number:")
    const dcert = getByLabelText("Dam certificate number:")

    fireEvent.change(id, {target:{value:"1"}})
    fireEvent.change(name, {target:{value:"test"}})
    fireEvent.change(scert, {target:{value:"2"}})
    fireEvent.change(dcert, {target:{value:"3"}})

    expect(id.value).toMatch("1")
    expect(name.value).toMatch("test")
    expect(scert.value).toMatch("2")
    expect(dcert.value).toMatch("3")



    //close the modal
    const closebtn = getByTestId('closebtn')
    fireEvent.click(closebtn)
    expect(mockOnClick).toHaveBeenCalledTimes(1)
    fireEvent.click(clickIndicator)
    expect(mockOnClick).toHaveBeenCalledTimes(1)

})


