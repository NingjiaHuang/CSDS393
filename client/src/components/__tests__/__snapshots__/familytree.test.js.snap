// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot of display tree 1`] = `
<div
  data-testid="displaytree-1"
>
   
  <input
    className="form-control"
    data-testid="searchTree"
    onChange={[Function]}
    placeholder="search a family member here"
    type="text"
  />
  <table
    class="table mt-5 text-center"
  >
    <thead>
      <tr>
        <th>
          Cat Name
        </th>
        <th>
          Certificate number
        </th>
        <th>
          Sire path
        </th>
        <th>
          Dam path
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`matches snapshot of family tree of breeder 1`] = `
<div
  className="container"
  data-test="container"
>
  <div
    className="mx-auto mt-5"
  >
    <h2
      className="text-center"
    >
      Family Tree
    </h2>
    <div>
      <button
        className="btn btn-info"
        data-testid="pairBtn"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Pair
      </button>
    </div>
    <div
      className="mt-3"
    >
      <div
        data-testid="displaytree-1"
      >
         
        <input
          className="form-control"
          data-testid="searchTree"
          onChange={[Function]}
          placeholder="search a family member here"
          type="text"
        />
        <table
          class="table mt-5 text-center"
        >
          <thead>
            <tr>
              <th>
                Cat Name
              </th>
              <th>
                Certificate number
              </th>
              <th>
                Sire path
              </th>
              <th>
                Dam path
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`matches snapshot of family tree of parents 1`] = `
<div
  className="container"
  data-test="container"
>
  <div
    className="mx-auto mt-5"
  >
    <h2
      className="text-center"
    >
      Family Tree
    </h2>
    <div
      className="mt-3"
    >
      <div
        data-testid="displaytree-1"
      >
         
        <input
          className="form-control"
          data-testid="searchTree"
          onChange={[Function]}
          placeholder="search a family member here"
          type="text"
        />
        <table
          class="table mt-5 text-center"
        >
          <thead>
            <tr>
              <th>
                Cat Name
              </th>
              <th>
                Certificate number
              </th>
              <th>
                Sire path
              </th>
              <th>
                Dam path
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`matches snapshot of pair cats 1`] = `
<div>
  <button
    className="btn btn-info"
    data-testid="pairBtn"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Pair
  </button>
</div>
`;
